plugins/inputs/elasticsearch_query/aggregation_query.go

```diff
@@ -0,0 +1,203 @@
+package elasticsearch_query
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	elastic "gopkg.in/olivere/elastic.v5"
+)
+
+type aggKey struct {
+	measurement string
+	name        string
+	function    string
+	field       string
+}
+
+type aggregationQueryData struct {
+	aggKey
+	isParent    bool
+	aggregation elastic.Aggregation
+}
+
+func (e *ElasticsearchQuery) runAggregationQuery(ctx context.Context, aggregation esAggregation) (*elastic.SearchResult, error) {
+
+	now := time.Now().UTC()
+	from := now.Add(aggregation.QueryPeriod.Duration * -1)
+	filterQuery := aggregation.FilterQuery
+
+	if filterQuery == "" {
+		filterQuery = "*"
+	}
+
+	query := elastic.NewBoolQuery()
+	query = query.Filter(elastic.NewQueryStringQuery(filterQuery))
+	query = query.Filter(elastic.NewRangeQuery(aggregation.DateField).From(from).To(now))
+
+	search := e.esClient.Search().Index(aggregation.Index).Query(query).Size(0)
+
+	// add only parent elastic.Aggregations to the search request, all the rest are subaggregations of these
+	for _, v := range aggregation.aggregationQueryList {
+		if v.isParent && v.aggregation != nil {
+			search.Aggregation(v.aggKey.name, v.aggregation)
+		}
+	}
+	searchResult, err := search.Do(ctx)
+
+	if err != nil && searchResult != nil {
+		return searchResult, fmt.Errorf("%s - %s", searchResult.Error.Type, searchResult.Error.Reason)
+	}
+
+	return searchResult, err
+
+}
+
+// GetMetricFields function returns a map of fields and field types on Elasticsearch that matches field.MetricFields
+func (e *ElasticsearchQuery) getMetricFields(ctx context.Context, aggregation esAggregation) (map[string]string, error) {
+	mapMetricFields := make(map[string]string)
+
+	if e.esClient == nil {
+		err := e.connectToES()
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
Do we really need this here? If so you should also care for this in `runAggregationQuery`.